### PR TITLE
ES 5.2.2 Support

### PR DESCRIPTION
**Pull request by adrianocrestani**
- I dropped guava dependency. Added Memoizer class to replace depency on guava's Suppliers.memoize()
- bumped analyzer version from 2 to 3
- adapted AnalysisKoreanPlugin to the new ES Plugin API

### CODE REVIEW

File: src/main/java/org/apache/lucene/analysis/kr/utils/DictionaryUtil.java [hunk 1]
```diff
@@ -23,100 +23,71 @@
...
-            }
-
-            return Collections.unmodifiableMap(cjwords);
-        }
-    });
+	private static Supplier<Map<String, String>> JOSAS = Memoizer.memoize(() -> readDict(KoreanEnv.FILE_JOSA));
 	
+	private static Supplier<Map<String, String>> EOMIS = Memoizer.memoize(() -> readDict(KoreanEnv.FILE_EOMI));
+
+    private static Supplier<Map<String, String>> PREFIXS = Memoizer.memoize(() -> readDict(KoreanEnv.FILE_PREFIX));
```
Comment (jpdaigle) [inline src/main/java/org/apache/lucene/analysis/kr/utils/DictionaryUtil.java]: If this isn't part of the public API (it's private) we should fix spelling: `PREFIXES`

File: src/main/java/org/apache/lucene/analysis/kr/utils/DictionaryUtil.java [hunk 2]
```diff
@@ -23,100 +23,71 @@
...
+
+    private static Supplier<Map<String, String>> SUFFIXS = Memoizer.memoize(() -> readDict(KoreanEnv.FILE_SUFFIX));
```
Comment (jpdaigle) [inline src/main/java/org/apache/lucene/analysis/kr/utils/DictionaryUtil.java]: `SUFFIXES`

File: src/main/java/org/apache/lucene/analysis/kr/utils/DictionaryUtil.java [hunk 3]
```diff
@@ -23,100 +23,71 @@
...
+
+	private static Supplier<Map<String, WordEntry>> UNCOMPOUNDS = Memoizer.memoize(() -> {
+		Map<String, WordEntry> uncompounds = new HashMap<>();
+		List<String> lines;
+
+		try {
```
Comment (jpdaigle) [inline src/main/java/org/apache/lucene/analysis/kr/utils/DictionaryUtil.java:55]: This whole file reading thing is ugly and could be easily done in "vanilla" Java, but I know you just changed indentation and it's not your bug. 😐

File: src/main/java/org/apache/lucene/analysis/kr/utils/Memoizer.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+
+public class Memoizer {
+
+    public static <T> Supplier<T> memoize(Supplier<T> delegate) {
+        AtomicReference<T> value = new AtomicReference<>();
```
Comment (jpdaigle) [inline src/main/java/org/apache/lucene/analysis/kr/utils/Memoizer.java]: could be final

**Review by jpdaigle — CHANGES_REQUESTED** [review]
* Looks like the plugin init changed: what's our test plan for making sure this still loads and runs OK?

* If this is bumping the plugin's major version, do we want to keep this on a separate branch? Or better yet, branch off the ES-2.3 version to a maintenance branch and 5.2.2 goes to trunk?

File: src/main/java/org/apache/lucene/analysis/kr/utils/Memoizer.java (hunk 4 above)
Comment (adrianocrestani) [reply src/main/java/org/apache/lucene/analysis/kr/utils/Memoizer.java]: value or Memoizer class? Anyway, I will make both final

**Comment by adrianocrestani** [comment]
> Looks like the plugin init changed: what's our test plan for making sure this still loads and runs OK?

@jpdaigle , I have re-run the same tests I described [here](https://github.com/tripadvisor/elasticsearch-analysis-korean/pull/4) to make sure initialization still works.

>  Or better yet, branch off the ES-2.3 version to a maintenance branch and 5.2.2 goes to trunk?

That is exactly what I did :)

File: src/main/java/Test.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import java.io.StringReader;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * Created by acrestani on 2017-03-27.
+ */
+public class Test {
```
Comment (jpdaigle) [inline src/main/java/Test.java]: Name isn't very descriptive.

**Review by jpdaigle — APPROVED** [review]

File: src/main/java/Test.java (hunk 5 above)
Comment (adrianocrestani) [reply src/main/java/Test.java]: oh no, I shouldn't have committed that. I will revert it.

**Comment by jpdaigle** [comment]
:shipit:

**Pull request merged**